homeassistant/components/fan/template.py

```diff
@@ -0,0 +1,349 @@
+        return self._name
+
+    @property
+    def supported_features(self) -> int:
+        """Flag supported features."""
+        return self._supported_features
+
+    @property
+    def speed_list(self: ToggleEntity) -> list:
+        """Get the list of available speeds."""
+        return self._speed_list
+
+    @property
+    def is_on(self):
+        """Return true if device is on."""
+        return self._state
+
+    @property
+    def speed(self):
+        """Return the current speed."""
+        return self._speed
+
+    @property
+    def oscillating(self):
+        """Return the oscillation state."""
+        return self._oscillating
+
+    @property
+    def should_poll(self):
+        """Return the polling state."""
+        return False
+
+    @asyncio.coroutine
+    def async_turn_on(self, speed: str=None) -> None:
+        """Turn on the fan.
+
+        This method is a coroutine.
+        """
+        self._state = True
+        self.hass.async_add_job(self._on_script.async_run())
+
+        if speed:
+            yield from self.async_set_speed(speed)
+
+    @asyncio.coroutine
+    def async_turn_off(self) -> None:
+        """Turn off the fan.
+
+        This method is a coroutine.
+        """
+        self._state = False
+        self.hass.async_add_job(self._off_script.async_run())
+
+    @asyncio.coroutine
+    def async_set_speed(self, speed: str) -> None:
+        """Set the speed of the fan.
+
+        This method is a coroutine.
+        """
+        if self._set_speed_script is None:
+            return
+
+        if speed in self._speed_list:
+            self._speed = speed
+            self.hass.async_add_job(
+                self._set_speed_script.async_run({ATTR_SPEED: speed}))
+        else:
+            _LOGGER.error(
+                'Received invalid speed: %s. ' +
+                'Expected: %s.',
+                speed, self._speed_list)
+
+    @asyncio.coroutine
+    def async_oscillate(self, oscillating: bool) -> None:
+        """Set oscillation of the fan.
+
+        This method is a coroutine.
+        """
+        if self._set_oscillating_script is None:
+            return
+
+        if oscillating is True or oscillating is False:
+            self._oscillating = oscillating
+            self.hass.async_add_job(self._set_oscillating_script.async_run(
+                {ATTR_OSCILLATING: oscillating}))
+        else:
+            _LOGGER.error(
+                'Received invalid oscillating: %s. ' +
+                'Expected True/False.', oscillating)
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Register callbacks."""
+        @callback
+        def template_fan_state_listener(entity, old_state, new_state):
+            """Handle target device state changes."""
+            self.async_schedule_update_ha_state(True)
+
+        @callback
+        def template_fan_startup(event):
+            """Update template on startup."""
+            async_track_state_change(
```

Comment:
`self.hass.helpers.event.async_track...`

Reply:
I dont get what you mean here. I already imported `from homeassistant.helpers.event import async_track_state_change`

Reply:
You don't need import it, you can access with `self.hass.hepers.event.`

Reply:
done

Reply:
I am not sure why but there was error when using `self.hass.hepers.event.async_track_state_change`. I need to revert this back. And I checked in the repo, there is no place using `self.hass.hepers.event.async_track_state_change`

Reply:
Seems like you had a typo. It should be `self.hass.helpers.event.async_track_state_change`.

Reply:
You used it with incorrect arguments. When using `self.hass.helpers.event.async_track_state_change` you should not pass `hass` as first argument.
```py
self.hass.helpers.event.async_track_state_change(
    self._entities, template_fan_state_listener)
```

Reply:
Thank @MartinHjelmare